### ISSUE TITLE
correlation() modifies input matrix

### ISSUE THREAD

**Issue by zemlyansky** [mljs/matrix#102]
Similarly to #97 `correlation(X)` changes `X`. That makes sense with `abs` or other transformations, but what's the motivation behind mutating `X` when computing a statistic? Thanks!

```javascript
const { Matrix, correlation } = require('ml-matrix')
const matrix = new Matrix([[1,2,3], [4,3,6], [7,1,9]])
const corr = correlation(matrix)
console.log(matrix.to2DArray())
```
produces `[[-1, 0, -1], [0, 1, 0], [1, -1, 1]]`

**Comment by targos** [comment]
This is a bug indeed. The solution is to use `new Matrix()` instead of `Matrix.checkMatrix` in https://github.com/mljs/matrix/blob/master/src/correlation.js. The covariance function is also affected: https://github.com/mljs/matrix/blob/master/src/covariance.js

Would you like to open a pull request to fix it?

**Comment by zemlyansky** [comment]
Just tried to fix it locally changing 
`xMatrix = Matrix.checkMatrix(xMatrix)` to `xMatrix = new Matrix(xMatrix)` and
`yMatrix = Matrix.checkMatrix(yMatrix)` to `yMatrix = new Matrix(yMatrix)` 
but after that change tests fail with such message:
```
expect(received).toStrictEqual(expected) // deep equality

- Expected
+ Received

  Array [
-   -0.042727630263691335,
-   -0.042727630263691335,
-   1.0000000000000002,
-   -0.042727630263691335,
-   -0.042727630263691335,
+   -0.04272763026369126,
+   -0.04272763026369126,
+   1,
+   -0.04272763026369126,
+   -0.04272763026369126,
  ]

  14 |       [8, 16, 24, 32, 40],
  15 |     ]);
> 16 |     expect(Array.from(correlation(y).data[2])).toStrictEqual([
```
I can make that rule to be less strict rounding numbers, but it's not 100% clear why output actually changes

**Comment by targos** [comment]
Yeah, that's probably because of the bug. When only one parameter is passed to `correlation`, it is used twice (as xMatrix and yMatrix), so when the matrices are centered, x is centered correctly, but then it is centered again, because x and y are the same (but we forgot this case, so `yIsSame` is false in this case).

I think you can safely update the expected values in the test to the new ones.

**Comment by targos** [comment]
Thanks for the fix. It's released in v6.5.0.
